src/EncoderDecoder/Huffman/HuffmanEncoderDecoder.cpp

```diff
@@ -0,0 +1,252 @@
+#include "EncoderDecoder/Huffman/HuffmanEncoderDecoder.h"
+
+HuffmanEncoderDecoder::NodeVector HuffmanEncoderDecoder::getFrequencies(
+	const std::string &plaintext) {
+
+	// Initialize frequencies
+	NodeVector frequencies;
+	frequencies.reserve(256);
+	for(int c = 0; c < 256; c++) {
+		frequencies.push_back(std::make_unique<Node>(Node(c)));
+	}
+
+	// Count frequency of each char in plaintext
+	std::for_each(plaintext.cbegin(), plaintext.cend(), [&](unsigned char c) {
+		auto &fp = frequencies[c];
+		(*fp)++;
+	});
+
+	return frequencies;
+}
+
+HuffmanEncoderDecoder::NodePtr HuffmanEncoderDecoder::buildPrefixlessTree(
+	NodeVector &frequencies) {
+
+	// Sort by ascending order of frequency
+	std::sort(frequencies.begin(),
+		frequencies.end(),
+		[](const NodePtr &n1, const NodePtr &n2) {
+			return n1->frequency < n2->frequency;
+		});
+
+	// Find the first element with non-zero frequency
+	auto activeSymbols = std::upper_bound(
+		frequencies.begin(), frequencies.end(), 0, [](int value, NodePtr &n) {
+			return n->frequency > value;
+		});
+
+	// Copy to list non-zero frequency elements
+	std::list<NodePtr> frequenciesList;
+	std::copy(std::make_move_iterator(activeSymbols),
+		std::make_move_iterator(frequencies.end()),
+		std::back_inserter(frequenciesList));
+
+	// If there are no elements, return an empty node
+	if(frequenciesList.empty()) {
+		return nullptr;
+	}
+
+	// While there is more than one element in the list
+	while(frequenciesList.size() > 1) {
+		// Pop first 2 elements
+		auto n1 = std::move(frequenciesList.front());
+		frequenciesList.pop_front();
+		auto n2 = std::move(frequenciesList.front());
+		frequenciesList.pop_front();
+		// Create a new one which is their sum
+		auto newNode = n1 + n2;
+		// Insert it to the correct location in the list
+		auto newLoc = std::upper_bound(frequenciesList.begin(),
+			frequenciesList.end(),
+			newNode->frequency,
+			[](int frequency, NodePtr &n) { return n->frequency > frequency; });
+		frequenciesList.insert(newLoc, std::move(newNode));
+	}
+
+	return std::move(frequenciesList.front());
+}
+
+std::pair<bool, HuffmanEncoderDecoder::MappingType>
+	HuffmanEncoderDecoder::createMapping(const NodePtr &root) {
```

Comment:
The typedefs confuse me, is this a `const unique_ptr&`?

Reply:
Yes, why?
A note: This function generates a `MappingType` object from the given tree, and discards the tree, so this signature makes sense (at least to me).

Reply:
Passing a reference means there’s no ownership transfer, which makes the `unique_ptr` redundant and forces the user to use one, if you do not intend to transfer ownership to your function, use a raw pointer. If you do, you need the pointer by value, this way it will be moved inside the function

Reply:
You're right! Might be the alias confused me too! It should definitely pass by value.

TODO